### ISSUE TITLE
Interpreting the parser

### ISSUE THREAD

**Issue by TrentBrick** [aqlaboratory/proteinnet#7]
TLDR; How can I best use the parser provided to go from tf DataRecords to inputting this into a Keras model?

Preamble: Thank you very much for compiling this awesome dataset that I am excited to prototype some protein design ideas on! 

Disclaimer: I am quite new to Tensorflow. Below I outline every bump I ran into along the way of trying to work with this fantastic dataset. If this is going to be like ImageNet for proteins it needs to have more user friendly documentation on how to get up and running with it. I hope that the following will outline roadbumps for future users like myself. If the following is helpful I would love to help make this dataset more user friendly by providing example code etc. 

----------------

I have spent over 10 hours trying to figure out how to work with the Tensorflow records provided. I assumed this was a very easy data format to work with before learning that it was notorious for having poor documentation and being difficult to understand. 

A "parser" is provided but I spent quite some time assuming that it was, as per the tf documentation, a parser that could be applied in some way through dataset=tf.data.TFRecordDataset(files) and then dataset.map(read_protein). Instead it is a standalone function that you must first create a filequeue for. 

It was only after spending considerable time going through the RGN docs to understand what was going on and how to structure my import after finding the _dataflow and read_protein functions in this separate repo. Then, because the imports are tensors, I was unable to look inside them at the data actually being imported. Eager execution also does not work because we are using tf.data. 

Going through the RGN files I ended up reusing a lot of the _dataflow function in model.py. However, even this is hard to understand. I still do not know what the distinction between batch size and "num_steps" is after using tf.contrib.training.bucket_by_sequence_length and what the exact generator is that can then be given to fit_generator in order to train a keras model using this data. 

After all of this time I have decided to go back to square one by instead downloading the .txt files and working with those instead. That way I can at least get a basic model up and running before I try to work with more advanced DataRecords and build a more efficient pipeline with them.

**Comment by alquraishi** [comment]
Thanks for your kind words and your constructive feedback. You are quite right about the parser being difficult to use with `tf.data`. The code was written in low-level TF and since it worked for my purposes (and is very performant), I never had the need to reimplement using `tf.data`. This is partly why I made the text-based file formats available, so that people can easily build parsers on top of them.

If you feel inclined to write a parser compatible with `tf.data` I'd be happy to incorporate it into the codebase. If not I will try to write one, but it will be probably be some time before I get around to it.

**Comment by TrentBrick** [comment]
Thank you for your reply. And I am glad that my feedback was useful. I am happy to be a guinnea pig for making the ProteinNet database as accessible as possible! You have certainly saved me a **lot** of time by creating this in the first place. 

What's the best way for me to share any code which you can then decide whether or not to add to the codebase? Pull requests? 

I am going to work with the text files for now but foresee moving over to `tf.data` once I get my models up and working.

**Comment by alquraishi** [comment]
No problem! Pull requests are fine. And if I end up taking a stab at it sooner than I expected I'll post an update here.

**Comment by TrentBrick** [comment]
I have switched over to Pytorch so won't be making a nice Tensorflow parser anytime soon. However, I did stumble upon this which you may want to copy and cite (it has an MIT license) as a fast way to parse the text file, apply the masking to it, etc. https://github.com/OpenProtein/openprotein/blob/master/preprocessing.py

**Comment by alquraishi** [comment]
That's great! Thanks for bringing it to my attention. I added a link on the landing page to the code.

**Comment by v-i-s-h** [comment]
Hi,

How to just read data from the tf_records. I tried using the helper functions provided in `tf_parser.py` added the following code below it:
```
if __name__ == '__main__':
    filename_queue = tf.train.string_input_producer(["./../data/testing.tfrec"], num_epochs=None)

    data = read_protein(filename_queue, 2)
    
    # Printout the corresponding tensors
    for d in data:
        print("\t", d)    

    with tf.Session(config=tf.ConfigProto(allow_soft_placement=True, log_device_placement=False)) as sess:

        coord = tf.train.Coordinator()
        threads = tf.train.start_queue_runners(coord=coord, sess=sess)
        
        sess.run(tf.local_variables_initializer())
        
        d_ = sess.run(data)

        print(d_)
```
After printing out the regular messages given below, the script freezes and didn't print any output. I cannot even use Ctrl+C to kill the script. I have to kill it using task manager.
```
WARNING:tensorflow:From tf_parser.py:93: string_input_producer (from tensorflow.python.training.input) is deprecated and will be removed in a future version.
Instructions for updating:
Queue-based input pipelines have been replaced by `tf.data`. Use `tf.data.Dataset.from_tensor_slices(string_tensor).shuffle(tf.shape(input_tensor, out_type=tf.int64)[0]).repeat(num_epochs)`. If `shuffle=False`, omit the `.shuffle(...)`.
WARNING:tensorflow:From /opt/Miniconda2/envs/casp_py2/lib/python2.7/site-packages/tensorflow/python/training/input.py:278: input_producer (from tensorflow.python.training.input) is deprecated and will be removed in a future version.
Instructions for updating:
Queue-based input pipelines have been replaced by `tf.data`. Use `tf.data.Dataset.from_tensor_slices(input_tensor).shuffle(tf.shape(input_tensor, out_type=tf.int64)[0]).repeat(num_epochs)`. If `shuffle=False`, omit the `.shuffle(...)`.
WARNING:tensorflow:From /opt/Miniconda2/envs/casp_py2/lib/python2.7/site-packages/tensorflow/python/training/input.py:190: limit_epochs (from tensorflow.python.training.input) is deprecated and will be removed in a future version.
Instructions for updating:
Queue-based input pipelines have been replaced by `tf.data`. Use `tf.data.Dataset.from_tensors(tensor).repeat(num_epochs)`.
WARNING:tensorflow:From /opt/Miniconda2/envs/casp_py2/lib/python2.7/site-packages/tensorflow/python/training/input.py:199: __init__ (from tensorflow.python.training.queue_runner_impl) is deprecated and will be removed in a future version.
Instructions for updating:
To construct input pipelines, use the `tf.data` module.
WARNING:tensorflow:From /opt/Miniconda2/envs/casp_py2/lib/python2.7/site-packages/tensorflow/python/training/input.py:199: add_queue_runner (from tensorflow.python.training.queue_runner_impl) is deprecated and will be removed in a future version.
Instructions for updating:
To construct input pipelines, use the `tf.data` module.
WARNING:tensorflow:From /opt/Miniconda2/envs/casp_py2/lib/python2.7/site-packages/tensorflow/python/training/input.py:202: to_float (from tensorflow.python.ops.math_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.cast instead.
WARNING:tensorflow:From tf_parser.py:62: __init__ (from tensorflow.python.ops.io_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Queue-based input pipelines have been replaced by `tf.data`. Use `tf.data.TFRecordDataset`.
WARNING:tensorflow:From tf_parser.py:74: to_int32 (from tensorflow.python.ops.math_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.cast instead.
WARNING:tensorflow:From /opt/Miniconda2/envs/casp_py2/lib/python2.7/site-packages/tensorflow/python/ops/control_flow_ops.py:423: colocate_with (from tensorflow.python.framework.ops) is deprecated and will be removed in a future version.
Instructions for updating:
Colocations handled automatically by placer.
('\t', <tf.Tensor 'read_protein/strided_slice:0' shape=() dtype=string>)
('\t', <tf.Tensor 'read_protein/one_hot:0' shape=(?, 20) dtype=float32>)
('\t', <tf.Tensor 'read_protein/ParseSingleSequenceExample/ParseSingleSequenceExample:1' shape=(?, 21) dtype=float32>)
('\t', <tf.Tensor 'read_protein/ToInt32_1:0' shape=(?,) dtype=int32>)
('\t', <tf.Tensor 'read_protein/ParseSingleSequenceExample/ParseSingleSequenceExample:5' shape=(?, 3) dtype=float32>)
('\t', <tf.Tensor 'read_protein/ter_mask/mul_1:0' shape=(?, ?) dtype=float32>)
('\t', <tf.Tensor 'read_protein/Size:0' shape=() dtype=int32>)
('\t', <tf.Tensor 'read_protein/LessEqual:0' shape=() dtype=bool>)
2019-05-26 10:51:52.912787: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2019-05-26 10:51:52.933927: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 3600000000 Hz
2019-05-26 10:51:52.934551: I tensorflow/compiler/xla/service/service.cc:150] XLA service 0x562c12017530 executing computations on platform Host. Devices:
2019-05-26 10:51:52.934568: I tensorflow/compiler/xla/service/service.cc:158]   StreamExecutor device (0): <undefined>, <undefined>
WARNING:tensorflow:From tf_parser.py:103: start_queue_runners (from tensorflow.python.training.queue_runner_impl) is deprecated and will be removed in a future version.
Instructions for updating:
To construct input pipelines, use the `tf.data` module.
```
How can I just read the data from tfrecord files?

**Comment by alquraishi** [comment]
Hi,

If you're run just interested in reading ProteinNet records, I suggest using the text-based version as it's much easier to parse, and I have a [parser](https://github.com/aqlaboratory/proteinnet/blob/master/code/text_parser.py) for that. If you'd like to read the `TFRecord` format within TensorFlow, the RGN code, specifically [here](https://github.com/aqlaboratory/rgn/blob/master/model/model.py#L544), contains an example
